### PR TITLE
#1734 actions based vr input

### PR DESCRIPTION
**Pull request by richardTingle**
This leaves all the old stuff there (the oculus, osvr & other openvr) but focusses on the LWJGL openvr for the short term future (I guess OpenXR later on as the more long term future). I have ended up with a bit of a leaky abstraction where VRInputAPI wants to talk about openVr specific stuff like action manifests but I couldn't avoid that without requiring users to cast to LWJGLOpenVRInput which feels equally as bad (especially as LWJGLOpenVRInput is probably the only one that's going to be usable till OpenXR comes along)

This allows Jmonkey VR to use the actions manifest style VR (which is the non deprecated openVr api for mapping button presses to "stuff" in game).

An example action manifest might look like

```
{
  "default_bindings": [
    {
      "controller_type": "oculus_touch",
      "binding_url": "oculusTouchDefaults.json"
    }
  ],
  "actions": [
    {
      "name": "/actions/main/in/OpenInventory",
      "requirement": "mandatory",
      "type": "boolean"
    },
    {
      "name": "/actions/main/in/scroll",
      "type": "vector2",
      "requirement": "mandatory"
    },
    {
      "name": "/actions/main/out/hapticTest",
      "type": "vibration",
      "requirement": "optional"
    },
    {
      "name": "/actions/main/in/leftTrigger",
      "type": "vector1",
      "requirement": "mandatory"
    }
  ],
  "action_sets": [
    {
      "name": "/actions/main",
      "usage": "leftright"
    }
  ],
  "localization" : [
    {
      "language_tag": "en_us",
      "/actions/main" : "My Game Actions",
      "/actions/main/in/OpenInventory" : "Open Inventory"
    }
  ]
}
```

With a default bindings looking like

```
{
  "action_manifest_version" : 0,
  "bindings": {
    "/actions/main": {
      "haptics" : [
        {
          "output" : "/actions/main/out/hapticTest",
          "path" : "/user/hand/left/output/haptic"
        }
      ],
      "sources" : [
        {
          "inputs" : {
            "click" : {
              "output" : "/actions/main/in/OpenInventory"
            }
          },
          "mode" : "button",
          "path" : "/user/hand/left/input/x"
        },
        {
          "inputs" : {
            "position" : {
              "output" : "/actions/main/in/scroll"
            }
          },
          "mode" : "joystick",
          "path" : "/user/hand/left/input/joystick"
        },
        {
          "inputs" : {
            "pull" : {
              "output" : "/actions/main/in/leftTrigger"
            }
          },
          "mode" : "trigger",
          "path" : "/user/hand/left/input/trigger"
        }
      ]
    }
  },
  "category" : "steamvr_input",
  "controller_type" : "oculus_touch",
  "description" : "Bindings for a oculusTouch controller",
  "name" : "Bindings for a oculusTouch controller",
  "options" : {},
  "simulated_actions" : []
}
```

Then it would be used like:

```
    public static void main(String[] args) {
        AppSettings settings = new AppSettings(true);
        VREnvironment env = new VREnvironment(settings);
        env.initialize();

        if (env.isInitialized()){
            VRAppState vrAppState = new VRAppState(settings, env);
            vrAppState.getVRinput().registerActionManifest("C:/Users/richa/Documents/Development/android/jmonkeyVrTest/src/main/resources/actionManifest.json", "/actions/main");

            Main app = new Main(vrAppState);
            app.setLostFocusBehavior(LostFocusBehavior.Disabled);
            app.setSettings(settings);
            app.setShowSettings(false);
            app.start();
        }
    }
```

And within the update loop like

```
    List<Geometry> handGeometries = new ArrayList<>();

    @Override
    public void simpleUpdate(float tpf) {

        VRAppState vrAppState = getStateManager().getState(VRAppState.class);
        int numberOfControllers = vrAppState.getVRinput().getTrackedControllerCount(); //almost certainly 2, one for each hand

        //build as many geometries as hands, as markers for the demo (Will only tigger on first loop or if number of controllers changes)
        while(handGeometries.size()<numberOfControllers){
            Box b = new Box(0.1f, 0.1f, 0.1f);
            Geometry handMarker = new Geometry("hand", b);
            Material mat = new Material(assetManager, "Common/MatDefs/Misc/Unshaded.j3md");
            mat.setColor("Color", ColorRGBA.Red);
            handMarker.setMaterial(mat);
            rootNode.attachChild(handMarker);
            handGeometries.add(handMarker);
        }

        VRInputAPI vrInput = vrAppState.getVRinput();

        for(int i=0;i<numberOfControllers;i++){
            if (vrInput.isInputDeviceTracking(i)){ //might not be active currently, avoid NPE if that's the case

                Vector3f position = vrInput.getFinalObserverPosition(i);
                Quaternion rotation = vrInput.getFinalObserverRotation(i);

                Geometry geometry = handGeometries.get(i);
                geometry.setLocalTranslation(position);
                geometry.setLocalRotation(rotation);
            }
        }

        if (vrInput.getDigitalActionState("/actions/main/in/OpenInventory").state){
            System.out.println("openInventory");
            vrInput.triggerHapticAction("/actions/main/out/hapticTest", 1, 5, 1 );
        }

        System.out.println("left trigger" + vrInput.getAnalogActionState("/actions/main/in/leftTrigger").x);
    }
```

Assuming people are happy with this I intend to expand the wiki to document how to write action manifests etc

### CODE REVIEW

File: jme3-vr/src/main/java/com/jme3/input/vr/lwjgl_openvr/LWJGLOpenVRInput.java [hunk 1]
```diff
@@ -106,6 +163,87 @@ public LWJGLOpenVRInput(VREnvironment environment) {
...
+            int errorCode = VRInput.VRInput_GetActionHandle(actionSet, longBuffer);
+            if ( errorCode != 0 )
+            {
+                logger.warning( "An error code of " + errorCode + " was reported while fetching an action set handle for " + actionSet );
+            }
+            actionSetHandle = longBuffer.get(0);
+            actionSetHandles.put(actionSet,actionSetHandle);
+        }
+
+        //Todo: this seems to imply that you could have multiple active action sets at once (Although I was not able to get that to work), allow multiple action sets
```
Comment (richardTingle) [inline jme3-vr/src/main/java/com/jme3/input/vr/lwjgl_openvr/LWJGLOpenVRInput.java:197]: It annoyed me that I couldn't get multiple action sets working. But then I didn't really understand why you'd want more than one action set, or why you'd want to turn them off. So I haven't worried about it too much

Comment (grizeldi) [reply jme3-vr/src/main/java/com/jme3/input/vr/lwjgl_openvr/LWJGLOpenVRInput.java:206]: The way multiple action sets could be used is, for example, if you want to have input per AppState. Or maybe one action set for interacting with the world and another for interacting with the user interface. There definitely are use cases, but I don't think it's a big deal if we don't support it, especially as the OpenVR documentation is a bit vague on how that is supposed to work.

**Review by grizeldi — COMMENTED** [review]
Wrote down a couple of documentation nitpicks, but overall from a code perspective looks good to me. I don't have any jme powered VR projects on hand, nor the time to make one from scratch, so I couldn't test if it actually runs though.

File: jme3-vr/src/main/java/com/jme3/input/vr/VRInputAPI.java [hunk 2]
```diff
@@ -4,40 +4,178 @@
...
+
+    /**
+     * Registers an action manifest. An actions manifest is a file that defines "actions" a player can make.
+     * (An action is an abstract version of a button press). The action manifest may then also include references to
+     * further files that define default mappings between those actions and physical buttons on the VR controllers.
+     *
+     * Note that registering an actions manifest will deactivate legacy inputs (i.e. methods such as {@link #isButtonDown}
+     * will no longer work
+     *
+     * This option is only relevant to OpenVR
```
[the remark this answers is not in the document]
Comment (richardTingle) [reply jme3-vr/src/main/java/com/jme3/input/vr/VRInputAPI.java:27]: Ooo, that's very useful. I ended up backwards engineering how action manifests worked from the examples I found but thats much more useful, thanks. I've added that

File: jme3-vr/src/main/java/com/jme3/input/vr/VRInputAPI.java [hunk 3]
```diff
@@ -4,40 +4,178 @@
...
+     * @return the DigitalActionState that has details on if the state has changed, what the state is etc.
+     */
+    default AnalogActionState getAnalogActionState( String actionName, String restrictToInput ){
+        throw new UnsupportedOperationException("Action manifests are not supported for the currently used VR API");
+    }
+
     /**
      * Check if the given button is down (more generally if the given input type is activated).
+     *
+     * Deprecated as should use an actions manifest approach. See {@link #registerActionManifest}
```
[the remark this answers is not in the document]
Comment (richardTingle) [reply jme3-vr/src/main/java/com/jme3/input/vr/VRInputAPI.java]: Fair point. I've made that clear on the deprecation notices

File: jme3-vr/src/main/java/com/jme3/input/vr/VRInputAPI.java [hunk 4]
```diff
@@ -4,40 +4,178 @@
...
+     * holding a weapon a button may be bound to "eject magazine" to allow you to load a new one, but that would only
+     * want to take effect on the hand that is holding the weapon
+     *
+     * Note that restrictToInput only restricts, it must still be bound to the input you want to receive the input from in
+     * the action manifest default bindings.
+     *
+     * {@link #registerActionManifest} must have been called before using this method.
+     *
+     * @param actionName The name of the action. E.g. /actions/main/in/openInventory
+     * @param restrictToInput the input to restrict the action to. E.g. /user/hand/right. Or null, which means "any input"
```
[the remark this answers is not in the document]
Comment (richardTingle) [reply jme3-vr/src/main/java/com/jme3/input/vr/VRInputAPI.java:91]: Sure, done

**Review by grizeldi — APPROVED** [review]
Looks good to me.

**Comment by stephengold** [comment]
Thanks for the review, @grizeldi.
> I couldn't test if it actually runs though.

@richardTingle Have you tested the new code in an application?

**Comment by richardTingle** [comment]
> Thanks for the review, @grizeldi.
> 
> > I couldn't test if it actually runs though.
> 
> @richardTingle Have you tested the new code in an application?

@stephengold yes, I did an install to maven local, created a project (with code much like in my intro to this PR) in intelliJ and ran on an oculus quest 2 (via virtual desktop so it pretended to be a PCVR system)

**Comment by stephengold** [comment]
Sounds good, then. Thanks for your contributions to the project, @richardTingle !

**Pull request merged**